### PR TITLE
Non memcmp slice comparison optimization

### PR DESCRIPTION
**Pull request by krtab**
This PR:

1. Changes the implementation for `[T]` slice comparison when `T: !BytewiseEq`. The previous implementation used `zip` which is not as optimized as it could be. Performance improvements are for example 20% when testing that `[Some(0_u64); 4096].as_slice() == [Some(0_u64); 4096].as_slice()`.
2. On musl (and maybe others), `memcmp` is equivalent to the new code added in the previous point. On such platforms it is better to never call `memcmp` because the compiler can leverage `T`'s alignment to generate better code than for `memcmp`'s `*const u8`.

I only realized when running `./x.py test` that platform specific code is not allowed in `core`. However, I would argue that this may be acceptable because:
- It does not affect portability so much, both implementations work fine, and this is only a matter of performance
- Deferring to libc's `memcmp` on the ground that it is more optimized than our generic one is already a kind of platform specific assumption. Arguably, rather than a `cfg(not()` making it opt-out, using `memcmp` should be opt-in.

This is likely more a draft than something ready to merge given the discussions that it must go through, but here is to starting these discussions.

### CODE REVIEW

**Comment by krtab** [comment]
This is loosely related to #113576 so I'm cross linking these two issues for the record.

File: library/core/src/slice/cmp.rs [hunk 1]
```diff
@@ -92,20 +91,28 @@ where
...
-impl<A, B> SlicePartialEq<B> for [A]
-where
-    A: BytewiseEq<B>,
-{
-    fn equal(&self, other: &[B]) -> bool {
-        if self.len() != other.len() {
-            return false;
-        }
+// and `memcmp` is not equivalent to our generic case
+#[cfg(not(target_env = "musl"))]
```
Comment (Nilstrieb) [inline library/core/src/slice/cmp.rs:95]: that cfg needs a clear comment (saying that musl's memcmp is not more optimized)

**Comment by Nilstrieb** [comment]
Let's run compiler benchmarks here too @bors try @rust-timer queue

File: library/core/src/slice/cmp.rs (hunk 1 above)
Comment (krtab) [reply library/core/src/slice/cmp.rs:95]: Thx for reviewing.
I'm planning on commenting more here and in `pal.rs` but I'm waiting to see how the debate on cfg in `core` goes.

**Comment by krtab** [comment]
I am willing to provide more benchmarks, especially on musl if needed, but I'd like to be sure that this won't be turned down on the ground of "no arch specific code in core" and before any performance consideration before doing so. Let me know what you feel is lacking for this to move forward.

**Comment by krtab** [comment]
#114382 makes the musl part of this PR irrelevant. To clarify the process, the `T: !BytewiseEq` is moved to its own PR: #116846.

**Pull request closed without merging**